File: tests/analytics/caffeine-sleep.test.ts
import { describe, test, expect } from 'vitest';
import { computeCaffeineSleepCutoff } from '$lib/analytics/caffeine-sleep';

// localHourTs encodes the hour in the runtime-local timezone (via setHours), so
// reading it back through the function with the same runtime tz round-trips
// regardless of where the test runs.
const TZ = Intl.DateTimeFormat().resolvedOptions().timeZone;

function makeSleepData(date: string, quality: number, duration: number) {
	return { date, sleepQuality: quality, sleepDurationMinutes: duration };
}

function localHourTs(date: string, localHour: number): string {
	const d = new Date(`${date}T00:00:00`);
	d.setHours(localHour);
	return d.toISOString();
}

describe('computeCaffeineSleepCutoff', () => {
	test('late caffeine with a clear, repeated sleep penalty yields a personal cutoff', () => {
		// 8 early days (last dose 08:00–11:00) followed by good sleep, 8 late days
		// (18:00–21:00) followed by poor sleep. The split has ≥5 nights a side and
		// survives the Bonferroni correction over the candidates tested.
		const caffeineEntries: { date: string; eatenAt: string; caffeine: number }[] = [];
		const sleepData: ReturnType<typeof makeSleepData>[] = [];
		for (let i = 0; i < 16; i++) {
			const day = `2024-01-${String(i + 1).padStart(2, '0')}`;
			const next = `2024-01-${String(i + 2).padStart(2, '0')}`;
			const late = i % 2 === 1;
			caffeineEntries.push({
				date: day,
				eatenAt: localHourTs(day, late ? 18 + (i % 4) : 8 + (i % 4)),
				caffeine: 100
			});
			sleepData.push(
				makeSleepData(next, late ? 3 + (i % 3) * 0.3 : 8 - (i % 3) * 0.3, late ? 300 : 480)
			);
		}
		const result = computeCaffeineSleepCutoff(caffeineEntries, sleepData, TZ);
		expect(result.estimatedCutoffHour).not.toBeNull();
		expect(result.estimatedCutoffHour!).toBeGreaterThanOrEqual(12);
		expect(result.estimatedCutoffHour!).toBeLessThanOrEqual(18);
		expect(result.comparisons).toBeGreaterThan(0);
		expect(result.pValue!).toBeLessThan(0.05);
		expect(result.defaultCutoffHour).toBe(14);
	});

	test('three nights a side is not enough to override the literature default', () => {
		const caffeineEntries = [
			{ date: '2024-01-01', eatenAt: localHourTs('2024-01-01', 19), caffeine: 100 },
			{ date: '2024-01-02', eatenAt: localHourTs('2024-01-02', 20), caffeine: 100 },
			{ date: '2024-01-03', eatenAt: localHourTs('2024-01-03', 21), caffeine: 100 },
			{ date: '2024-01-07', eatenAt: localHourTs('2024-01-07', 8), caffeine: 100 },
			{ date: '2024-01-08', eatenAt: localHourTs('2024-01-08', 9), caffeine: 100 },
			{ date: '2024-01-09', eatenAt: localHourTs('2024-01-09', 10), caffeine: 100 }
		];
		const sleepData = [
			makeSleepData('2024-01-02', 3, 300),
			makeSleepData('2024-01-03', 2, 280),
			makeSleepData('2024-01-04', 3, 290),
			makeSleepData('2024-01-08', 8, 480),
			makeSleepData('2024-01-09', 9, 500),
			makeSleepData('2024-01-10', 8, 490)
		];
		const result = computeCaffeineSleepCutoff(caffeineEntries, sleepData, TZ);
		expect(result.estimatedCutoffHour).toBeNull();
		expect(result.comparisons).toBe(0);
		expect(result.defaultCutoffHour).toBe(14);
	});

	test('early caffeine with no sleep impact gives null cutoff', () => {
		const caffeineEntries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T08:00:00+00:00', caffeine: 100 },
			{ date: '2024-01-02', eatenAt: '2024-01-02T09:00:00+00:00', caffeine: 100 },
			{ date: '2024-01-03', eatenAt: '2024-01-03T10:00:00+00:00', caffeine: 100 }
		];
		const sleepData = [
			makeSleepData('2024-01-02', 8, 480),
			makeSleepData('2024-01-03', 8, 480),
			makeSleepData('2024-01-04', 8, 480)
		];
		const result = computeCaffeineSleepCutoff(caffeineEntries, sleepData, TZ);
		expect(result.estimatedCutoffHour).toBeNull();
	});

	test('no caffeine data returns null cutoff and empty hourly impact', () => {
		const sleepData = [makeSleepData('2024-01-02', 8, 480), makeSleepData('2024-01-03', 7, 450)];
		const result = computeCaffeineSleepCutoff([], sleepData, TZ);
		expect(result.estimatedCutoffHour).toBeNull();
		expect(result.hourlyImpact).toHaveLength(0);
		expect(result.sampleSize).toBe(0);
	});

	test('hourlyImpact groups by last caffeine hour', () => {
		const hour = 14;
		// caffeine on day D → sleep on day D+1
		const caffeineEntries = [
			{ date: '2024-01-01', eatenAt: localHourTs('2024-01-01', hour), caffeine: 100 },
			{ date: '2024-01-02', eatenAt: localHourTs('2024-01-02', hour), caffeine: 100 }
		];
		const sleepData = [makeSleepData('2024-01-02', 7, 420), makeSleepData('2024-01-03', 7, 420)];
		const result = computeCaffeineSleepCutoff(caffeineEntries, sleepData, TZ);
		expect(result.hourlyImpact).toHaveLength(1);
		expect(result.hourlyImpact[0].hour).toBe(hour);
		expect(result.hourlyImpact[0].count).toBe(2);
	});

	test('skips caffeine entries without eatenAt', () => {
		const caffeineEntries = [
			{ date: '2024-01-01', eatenAt: null, caffeine: 100 },
			{ date: '2024-01-02', eatenAt: localHourTs('2024-01-02', 8), caffeine: 100 }
		];
		const sleepData = [makeSleepData('2024-01-03', 8, 480)];
		const result = computeCaffeineSleepCutoff(caffeineEntries, sleepData, TZ);
		expect(result.hourlyImpact).toHaveLength(1);
	});

	test('confidence reflects sample size', () => {
		const result = computeCaffeineSleepCutoff([], [], TZ);
		expect(result.confidence).toBe('insufficient');
	});
});

File: mobile/shared/src/commonTest/kotlin/com/bissbilanz/analytics/CaffeineSleepTest.kt
package com.bissbilanz.analytics

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

private fun pad2(n: Int): String = n.toString().padStart(2, '0')

class CaffeineSleepTest {
    @Test
    fun emptyDataReturnsEmpty() {
        val result = computeCaffeineSleepCutoff(emptyList(), emptyList())
        assertNull(result.estimatedCutoffHour)
        assertEquals(0, result.sampleSize)
        assertEquals(emptyList(), result.hourlyImpact)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }

    @Test
    fun hourlyBucketingTracksLastHourPerDay() {
        val caffeineEntries =
            listOf(
                CaffeineEntry(date = "2024-01-01", eatenAt = "2024-01-01T08:00:00Z", caffeine = 100.0),
                CaffeineEntry(date = "2024-01-01", eatenAt = "2024-01-01T14:00:00Z", caffeine = 80.0),
                CaffeineEntry(date = "2024-01-02", eatenAt = "2024-01-02T09:00:00Z", caffeine = 100.0),
            )
        val sleepData =
            listOf(
                SleepDataPoint(date = "2024-01-02", sleepQuality = 7.0, sleepDurationMinutes = 420.0),
                SleepDataPoint(date = "2024-01-03", sleepQuality = 8.0, sleepDurationMinutes = 450.0),
            )
        val result = computeCaffeineSleepCutoff(caffeineEntries, sleepData)
        // 2024-01-01 last caffeine at hour 14, sleep on 2024-01-02
        // 2024-01-02 last caffeine at hour 9, sleep on 2024-01-03
        assertEquals(2, result.hourlyImpact.size)
        val hour14 = result.hourlyImpact.find { it.hour == 14 }
        assertNotNull(hour14)
        assertEquals(7.0, hour14.avgQuality)
        val hour9 = result.hourlyImpact.find { it.hour == 9 }
        assertNotNull(hour9)
        assertEquals(8.0, hour9.avgQuality)
    }

    @Test
    fun cutoffDetectedWhenClearQualityDifference() {
        // Hours 8-11 (before cutoff 12): high quality sleep
        // Hours 16-19 (after cutoff 12): low quality sleep
        val caffeineEntries = mutableListOf<CaffeineEntry>()
        val sleepData = mutableListOf<SleepDataPoint>()
        val baseDate = "2024-01"
        // 5 days with early caffeine (hour 8-11) -> good sleep
        for (i in 1..5) {
            val date = "$baseDate-${pad2(i)}"
            val nextDate = "$baseDate-${pad2(i + 1)}"
            caffeineEntries.add(CaffeineEntry(date = date, eatenAt = "${date}T08:00:00Z", caffeine = 100.0))
            sleepData.add(SleepDataPoint(date = nextDate, sleepQuality = 9.0, sleepDurationMinutes = 480.0))
        }
        // 5 days with late caffeine (hour 18-19) -> poor sleep
        for (i in 10..14) {
            val date = "$baseDate-${pad2(i)}"
            val nextDate = "$baseDate-${pad2(i + 1)}"
            caffeineEntries.add(CaffeineEntry(date = date, eatenAt = "${date}T18:00:00Z", caffeine = 100.0))
            sleepData.add(SleepDataPoint(date = nextDate, sleepQuality = 4.0, sleepDurationMinutes = 360.0))
        }
        val result = computeCaffeineSleepCutoff(caffeineEntries, sleepData)
        assertNotNull(result.estimatedCutoffHour)
    }

    @Test
    fun missingSleepdataIsSkipped() {
        val caffeineEntries =
            listOf(
                CaffeineEntry(date = "2024-01-01", eatenAt = "2024-01-01T14:00:00Z", caffeine = 100.0),
            )
        // No sleep data for 2024-01-02 (the next date)
        val sleepData =
            listOf(
                SleepDataPoint(date = "2024-01-05", sleepQuality = 8.0, sleepDurationMinutes = 450.0),
            )
        val result = computeCaffeineSleepCutoff(caffeineEntries, sleepData)
        assertEquals(0, result.sampleSize)
        assertEquals(emptyList(), result.hourlyImpact)
    }

    @Test
    fun singleEntryReturnsSampleSizeOne() {
        val caffeine = listOf(CaffeineEntry("2024-01-01", "2024-01-01T14:00:00Z", 100.0))
        val sleep = listOf(SleepDataPoint("2024-01-02", 7.0, 420.0))
        val result = computeCaffeineSleepCutoff(caffeine, sleep)
        assertEquals(1, result.sampleSize)
        assertNull(result.estimatedCutoffHour)
    }

    @Test
    fun nullEatenAtEntriesSkipped() {
        val caffeine =
            listOf(
                CaffeineEntry("2024-01-01", null, 100.0),
                CaffeineEntry("2024-01-02", null, 200.0),
            )
        val sleep =
            listOf(
                SleepDataPoint("2024-01-02", 7.0, 420.0),
                SleepDataPoint("2024-01-03", 8.0, 480.0),
            )
        val result = computeCaffeineSleepCutoff(caffeine, sleep)
        assertEquals(0, result.sampleSize)
        assertEquals(emptyList(), result.hourlyImpact)
    }

    @Test
    fun zeroCaffeineEntriesSkipped() {
        val caffeine = listOf(CaffeineEntry("2024-01-01", "2024-01-01T14:00:00Z", 0.0))
        val sleep = listOf(SleepDataPoint("2024-01-02", 7.0, 420.0))
        val result = computeCaffeineSleepCutoff(caffeine, sleep)
        assertEquals(0, result.sampleSize)
    }

    @Test
    fun negativeCaffeineEntriesSkipped() {
        val caffeine = listOf(CaffeineEntry("2024-01-01", "2024-01-01T14:00:00Z", -50.0))
        val sleep = listOf(SleepDataPoint("2024-01-02", 7.0, 420.0))
        val result = computeCaffeineSleepCutoff(caffeine, sleep)
        assertEquals(0, result.sampleSize)
    }

    @Test
    fun nullQualityOnlySleepSkipped() {
        val caffeine = listOf(CaffeineEntry("2024-01-01", "2024-01-01T14:00:00Z", 100.0))
        val sleep = listOf(SleepDataPoint("2024-01-02", null, 420.0))
        val result = computeCaffeineSleepCutoff(caffeine, sleep)
        assertEquals(0, result.sampleSize)
    }

    @Test
    fun nullDurationOnlySleepSkipped() {
        val caffeine = listOf(CaffeineEntry("2024-01-01", "2024-01-01T14:00:00Z", 100.0))
        val sleep = listOf(SleepDataPoint("2024-01-02", 7.0, null))
        val result = computeCaffeineSleepCutoff(caffeine, sleep)
        assertEquals(0, result.sampleSize)
    }
}
